src/app/vertextool/qgsvertextool.cpp

```diff
@@ -2228,37 +2228,56 @@ void QgsVertexTool::moveVertex( const QgsPointXY &mapPoint, const QgsPointLocato
     addExtraSegmentsToEdits( edits, mapPoint, dragLayer, layerPoint );
   }
 
+  VertexEdits beforeEdits;
+  VertexEdits::const_iterator editLayer = edits.constBegin();
+  while ( editLayer != edits.constEnd() )
+  {
+    QgsVectorLayer *layer = editLayer.key();
+    QHash<QgsFeatureId, QgsGeometry>::const_iterator editGeometry = editLayer.value().constBegin();
+    while ( editGeometry != editLayer.value().constEnd() )
+    {
+      QgsFeatureId fid = editGeometry.key();
+      beforeEdits[layer][fid] = layer->getGeometry( fid );
+      ++editGeometry;
+    }
+    ++editLayer;
+  }
+
   applyEditsToLayers( edits );
 
   if ( QgsProject::instance()->topologicalEditing() )
   {
-    // topo editing: add vertex to existing segments when moving/adding a vertex to such segment.
-    // this requires that the snapping match is to a segment and the segment layer's CRS
-    // is the same (otherwise we would need to reproject the point and it will not be coincident)
     const auto editKeys = edits.keys();
     for ( QgsVectorLayer *layer : editKeys )
     {
-      const auto editGeom = edits[layer].values();
-      for ( QgsGeometry g : editGeom )
+      const auto editFid = edits[layer].keys();
+
+      for ( QgsFeatureId fid : editFid )
       {
-        QgsGeometry p = QgsGeometry::fromPointXY( QgsPointXY( layerPoint.x(), layerPoint.y() ) );
-        if ( ( ( mapPointMatch->hasEdge() || mapPointMatch->hasMiddleSegment() ) && mapPointMatch->layer() && layer->crs() == mapPointMatch->layer()->crs() )
-             || ( mapPointMatch->hasVertex() && !mapPointMatch->layer() && layer->crs() == mCanvas->mapSettings().destinationCrs() ) ) // also add topological points when snapped on intersection
-        {
-          if ( g.convertToType( QgsWkbTypes::PointGeometry, true ).contains( p ) )
-          {
-            if ( !layerPoint.is3D() )
-              layerPoint.addZValue( defaultZValue() );
-            layer->addTopologicalPoints( layerPoint );
-            if ( mapPointMatch->layer() )
-              mapPointMatch->layer()->addTopologicalPoints( layerPoint );
-          }
-        }
-        if ( QgsProject::instance()->avoidIntersectionsMode() != Qgis::AvoidIntersectionsMode::AllowIntersections )
+        const QgsGeometry newGeometry = edits[layer][fid];
+        const QgsGeometry newPointGeometry = newGeometry.convertToType( QgsWkbTypes::PointGeometry, true );
+        const QgsGeometry oldGeometry = beforeEdits[layer][fid];
+        const QgsGeometry oldPointGeometry = oldGeometry.convertToType( QgsWkbTypes::PointGeometry, true );
+
+        // Retrieves all modified vertices by all edits
+        const QgsGeometry editedPointsGeom = newPointGeometry.difference( oldPointGeometry );
```

Comment:
I'm a bit worried that this copying of geometries + executing difference to get the modified vertices can be expensive with big data.

Can we not use `mSelectedVertices` for that? It should already contain the vertices that we're modifying, no?

Reply:
The performance is much better, even with geometries with many vertices.

We can't use `mSelectedVertices` because if the user clicks directly on a vertex to move it, the list will be empty. Moreover, it will not contain the new vertices added by `avoidIntersection`.

Reply:
It's actually way slower on a dataset without a spatial index.

If you're moving a single vertex, it should be in `mDraggingVertex` instead.
`avoidIntersection` could take care of its own topological candidates and perform a diff only if required. Right now we're diffing potentially millions of identical points just because we moved a single vertex.

I know that the edited data should not lack a spatial index, but right now it works ok and there can be situations where the average user will not know the index is missing, like editing pasted data on a new scratch layer.